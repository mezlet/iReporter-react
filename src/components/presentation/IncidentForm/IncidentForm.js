/* eslint-disable jsx-a11y/label-has-associated-control */
/* eslint-disable jsx-a11y/label-has-for */

import React, { Component } from 'react';
import jsonForm from 'formdata-json';
import PropTypes from 'prop-types';
import { Form, Button, Image } from 'semantic-ui-react';

class IncidentForm extends Component {
  state = {
    imagePreviewUrl: null
  };

  handleImageChange = e => {
    e.preventDefault();

    const reader = new FileReader();
    const file = e.target.files[0];

    reader.addEventListener('load', () => {
      this.setState({
        imagePreviewUrl: reader.result
      });
    });

    reader.readAsDataURL(file);
  };

  handleSubmit = event => {
    event.preventDefault();
    const data = jsonForm(new FormData(event.target));

    const { createRecord, updateRecord } = this.props;

    if (data.id) {
      updateRecord(data);
    } else {
      createRecord(data);
    }
  };

  render() {
    const { imagePreviewUrl } = this.state;
    const {
      incident: { incident }
    } = this.props;
    return (
      <Form className="record-form" onSubmit={this.handleSubmit}>
        <Form.Field>
          <label>Title</label>
          <input
            type="text"
            name="title"
            defaultValue={incident.title}
            onChange={this.handleChange}
          />
        </Form.Field>
        <Form.Field>
          <label>Category</label>
          <select name="type">
            <option value={incident.type}>{incident.type}</option>
            <option value="red-flag">Redflag</option>
            <option value="intervention">Intervention</option>
          </select>
        </Form.Field>
        <Form.Field>
          <label>Location</label>
          <input type="text" name="location" defaultValue={incident.location} />
        </Form.Field>

        <Form.Field>
          <label>Description</label>
          <textarea
            rows="15"
            placeholder="Tell your story..."
            name="comment"
            defaultValue={incident.comment}
          />
        </Form.Field>
        <Form.Field className="upload-btn-wrapper">
          <label>Image</label>
          <Button className="form-btn">Upload Image</Button>
          <Image src={imagePreviewUrl || incident.images} />
          <input type="file" name="image" onChange={this.handleImageChange} />
        </Form.Field>
        <input type="hidden" name="id" defaultValue={incident.id} />
        <br />
        <Button color="black">Submit</Button>
      </Form>
    );
  }
}
IncidentForm.propTypes = {
  incident: PropTypes.shape({
    incident: PropTypes.shape()
  }),
  createRecord: PropTypes.func.isRequired,
  updateRecord: PropTypes.func.isRequired
};

export default IncidentForm;
