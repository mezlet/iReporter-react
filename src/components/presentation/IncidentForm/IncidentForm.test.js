/* eslint-disable no-unused-vars */
import React from 'react';
import { mount } from 'enzyme';
import IncidentForm from './IncidentForm';

const props = {
  createRecord: jest.fn(),
  updateRecord: jest.fn(),
  incident: {
    incident: {
      id: 1,
      type: 'sometext',
      title: 'anothertxt',
      location: 'here',
      createdon: 12343422222
    }
  }
};

const wrapper = mount(<IncidentForm {...props} />);

describe('<IncidentForm/> rendering', () => {
  it('should render without crashing', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('image changes are handled correctly', () => {
    const instance = wrapper.find('IncidentForm').instance();
    const fileInput = wrapper.find('input[name="image"]');

    const fileContents = 'file contents';
    const file = new Blob([fileContents], { type: 'text/plain' });
    const addEventListener = jest.fn((_, evtHandler) => {
      evtHandler();
    });
    const readAsDataURL = jest.fn();
    const dummyFileReader = {
      addEventListener,
      readAsDataURL,
      result: fileContents
    };
    window.FileReader = jest.fn(() => dummyFileReader);

    const event = {
      target: {
        files: [file]
      },
      preventDefault: jest.fn()
    };

    fileInput.simulate('change', event);

    expect(event.preventDefault).toHaveBeenCalled();
    expect(instance.state.imagePreviewUrl).toEqual(fileContents);
  });

  describe('handleSubmit', () => {
    const event = {
      preventDefault: jest.fn()
    };

    it('should handleSubmit', () => {
      const form = wrapper.find('Form');
      form.simulate('submit', event);

      expect(event.preventDefault).toHaveBeenCalled();
      expect(props.updateRecord).toHaveBeenCalled();
    });

    it('should handleSubmit', () => {
      props.incident.incident.id = null;
      const wrapperTwo = mount(<IncidentForm {...props} />);

      const form = wrapperTwo.find('Form');
      form.simulate('submit', event);

      expect(event.preventDefault).toHaveBeenCalledTimes(2);
      expect(props.createRecord).toHaveBeenCalled();
    });
  });
});
