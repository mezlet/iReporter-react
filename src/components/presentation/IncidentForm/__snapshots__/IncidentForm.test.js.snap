// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<IncidentForm/> rendering should render without crashing 1`] = `ReactWrapper {}`;
